ensime_shared/protocol_handler.py

```diff
@@ -0,0 +1,249 @@
+import webbrowser
+import json
+
+from ensime_shared.config import gconfig, feedback, commands
+from ensime_shared.util import catch
+from ensime_shared.symbol_format import completion_to_suggest
+
+class ProtocolHandler(object):
+
+    def __init__(self):
+        self.handlers = {}
+        self.register_responses_handlers()
+
+    def register_responses_handlers(self):
+        """Register handlers for responses from the server.
+
+        A handler must accept only one parameter: `payload`.
+        """
+        self.handlers["SymbolInfo"] = self.handle_symbol_info
+        self.handlers["IndexerReadyEvent"] = self.handle_indexer_ready
+        self.handlers["AnalyzerReadyEvent"] = self.handle_analyzer_ready
+        self.handlers["NewScalaNotesEvent"] = self.buffer_typechecks
+        self.handlers["BasicTypeInfo"] = self.show_type
+        self.handlers["ArrowTypeInfo"] = self.show_type
+        self.handlers["FullTypeCheckCompleteEvent"] = self.handle_typecheck_complete
+        self.handlers["StringResponse"] = self.handle_string_response
+        self.handlers["CompletionInfoList"] = self.handle_completion_info_list
+        self.handlers["TypeInspectInfo"] = self.handle_type_inspect
+        self.handlers["SymbolSearchResults"] = self.handle_symbol_search
+        self.handlers["DebugOutputEvent"] = self.handle_debug_output
+        self.handlers["DebugBreakEvent"] = self.handle_debug_break
+        self.handlers["DebugBacktrace"] = self.handle_debug_backtrace
+        self.handlers["DebugVmError"] = self.handle_debug_vm_error
+        self.handlers["RefactorDiffEffect"] = self.apply_refactor
+        self.handlers["ImportSuggestions"] = self.handle_import_suggestions
+        self.handlers["PackageInfo"] = self.handle_package_info
+
+    def handle_incoming_response(self, call_id, payload):
+        """Get a registered handler for a given response and execute it."""
+        self.log("handle_incoming_response: in {}".format(payload))
+        typehint = payload["typehint"]
+        handler = self.handlers.get(typehint)
+        def feature_not_supported(m):
+            msg = feedback["handler_not_implemented"]
+            self.raw_message(msg.format(typehint, self.launcher.ensime_version))
+        if handler:
+            with catch(NotImplementedError, feature_not_supported):
+                handler(call_id, payload)
```

Comment:
I don't agree, actually. The `catch` sugar is way more clear than a simple try-catch expression, since it clearly states what is doing, without the syntactic overhead of writing try catch all the time.